### ISSUE TITLE
[server-chart] cert-manager v0.6.0 dependency fails install without additional steps.

### ISSUE THREAD

**Issue by jgreat** [rancher/rancher#17700]
The cert-manager v0.6.0 helm chart now requires you to pre-install a static manifest for its CRDs. I feel that requiring additional steps outside the chart misses the point of using a helm chart. 

I've filed an upstream issue: https://github.com/helm/charts/issues/10949

**Why is this a problem?**
For most HA installs we rely on cert-manager to issue/generate ssl certs for Rancher.  Our current instructions which has the user just install cert-manager will end up erroring out.

What can we do?
In the short term we can update the HA install instructions to pin cert-manager version to v0.5.2.  

Longer term?
- Helm dependency management is a bit wonky, but we could copy the v0.5.2 version and add it as a static dependency to rancher server chart?
- Attempt to fix the upstream issue crd issue and submit it back to cert-manager.

**Comment by oskapt** [comment]
That issue saw no love from the helm maintainers and has been closed due to inactivity.

I saw today that the entire helm chart for cert-manager has been deprecated, and the chart refers people to https://github.com/jetstack/cert-manager/tree/master/deploy instead. 

Are we at risk of having the entire installation of cert-manager as listed in our docs one day fail because `cert-manager` fell out of the helm repo? 

The extra step, while annoying, isn't that big of a deal. We already have people creating service accounts for tiller, installing tiller, and so on. Running one more `kubectl` command to install the `cert-manager` CRDs is minor if it allows us to stay current with the latest release.

They're up to `0.8` now. We can't stay stuck on `0.5.2` forever when it's clear that they're not going to roll back the manual CRD step for us.

**Comment by oskapt** [comment]
I did a rancher deploy this weekend w/ RKE and cert-manager 0.8. I followed the jetstack instructions, and it worked beautifully. It's in the video that I created for doing an HA install of Rancher (to be released later this week). 

I propose that we just update the docs to include the extra steps. They do not add any significant overhead:

- install the CRDs
- use a namespace of `cert-manager`
- use the `jetstack` repo and chart

**Comment by dnauck** [comment]
> I did a rancher deploy this weekend w/ RKE and cert-manager 0.8. I followed the jetstack instructions, and it worked beautifully. It's in the video that I created for doing an HA install of Rancher (to be released later this week).
> 

How do you got this running?

See: https://github.com/rancher/rancher/issues/21242

**Comment by oskapt** [comment]
@dnauck - #21242 was rolled into #18535 and is about deploying helm charts with a relative URL from within the Rancher UI. it's not actually cert-manager specific. the latest comment says the issue will be resolved  in Rancher 2.3.

this issue is about not being able to use cert-manager > 0.5.2 to control certificate generation **_for the Rancher Server_** when installed into an RKE cluster. it works fine if you install the CRDs first, like the documentation for cert-manager says to do. 

the video will be out on Thursday.

**Comment by dnauck** [comment]
@oskapt yes, but how do you installed cert-manager with its new jetstack helm repository with rancher ui? (i ask because i get the error from the issue)

**Comment by oskapt** [comment]
An HA install of Rancher builds an RKE cluster first, then deploys cert-manager via `kubectl` and then installs Rancher via Helm. RKE is just Kubernetes, so cert-manager installs fine. The issue that you've encountered is a different issue from this one.

**Comment by Vashiru** [comment]
This issue just got a lot more pressing. I just got an e-mail from Let's encrypt stating that any versions older than 0.8.0 will be blocked starting November 1st, 2019. A copy from the e-mail can be seen below:

**ACTION REQUIRED: Let's Encrypt is blocking old cert-manager versions**

> We've been working with Jetstack, the authors of cert-manager, on a
> series of fixes to the client. Cert-manager sometimes falls into a
> traffic pattern where it sends really excessive traffic to Let's
> Encrypt's servers, continuously. To mitigate this, we plan to start
> blocking all traffic from cert-manager versions less than 0.8.0 (the
> current semver minor release), as of November 1, 2019. Please upgrade
> all of your cert-manager instances before then.
> 
> We're sending this email because this is the contact address of your
> cert-manager instance at:
> 
>  [stripped-ip-address] .
> 
> Version 0.8.0 is much better but we still observe excessive traffic in
> some cases. We're working with Jetstack to improve these cases. As new
> versions of cert-manager are released, we will add the non-current
> versions to our block list after 3 months. We strongly encourage
> cert-manager users to stay up-to-date with new versions.
> 
> Also, there is an opportunity to help both Jetstack and Let's Encrypt.
> Once you've upgraded, please check the logs for your cert-manager
> instances from time to time. Are they making excessive requests to Let's
> Encrypt (more than, say, 10 per day over multiple days)? If so, please
> share details at https://github.com/jetstack/cert-manager/issues/1948 .
> 
> Thanks,
> Let's Encrypt Team

@oskapt It might be helpful to have full instructions in the doc on how to uninstall the 0.5.2 version, install the CRDs and install the new one. And perhaps any caveats that might come with it.. (Like (I don't know if this is the case) if you deployed a certificate issuer with 0.5.2 you'll have to manually remove that first and reinstall once you've deployed 0.8.0).

**Comment by oskapt** [comment]
@Vashiru I made [a video](https://youtu.be/nF_k1YZYjH0) that covers this process step by step.

**Comment by oskapt** [comment]
There are pending changes for rancher/docs#1691 and rancher/charts#206 that will resolve this issue for both RKE/Rancher installations as well as installations from the App Catalog. Making this note so that they're all tied together.

**Comment by vkruoso** [comment]
Really looking forward to 2.3 release. Besides from the pressing about the version upgrade, it is really hard to get event v0.6+ running. The library chart also have issues on the current release. Due also to a bug in the catalog implementation (see #18535 ), we cannot use the jetstack helm chart (https://charts.jetstack.io) in Rancher to have an updated version. What is the timeframe on official release of 2.3?

**Comment by oskapt** [comment]
@vkruoso October 8

**Comment by Vashiru** [comment]
@oskapt Excellent video. It helped me upgrade our testing environment to the latest cert-manager. Only downside is it only discusses HTTP issuers and not DNS. And it appears (looking at the cert-manager docs) the documentation for DNS issuers hasn't been updated yet since it still uses `spec.acme`. For now I've resorted to HTTP-01 validation since I don't need wildcards at this time.

Am I correct though to assume that with the release of Rancher 2.3.0, a new chart for cert-manager will be released that allows me to simply upgrade from the 'Apps' section of Rancher? If that's the case I might prefer that rather than introducing a new dependency (Helm / Tiller) into our stack (added documentation, educating fellow developers, distributing certificates for the SSL setup).

**Comment by oskapt** [comment]
Thanks, @Vashiru! I'm glad it helped. It does only address HTTP, because the cert-manager docs have everything you need to add a DNS issuer. HTTP is universal, but DNS is specific to the provider that you want to use. Even if I made a video for one provider, it wouldn't solve everyone's needs.

Rancher 2.3 comes out next week. The cert-manager app update for the catalog is ready for merge, but it depends on a UI patch that will be in 2.3. You should be able to upgrade everything from the UI after next Tuesday.

**Comment by vkruoso** [comment]
Rancher 2.3 is out and Im trying it out now and it looks awesome! One problem related to using cert-manager app within Rancher is that it fails to handle helm charts that try to add resources to other namespaces. Im using the jetstack version of the helm chart, and get a ```Failed to install app cert-manager. Error: release cert-manager failed: resource's namespace kube-system doesn't match the current namespace cert-manager``` when trying to install the app in the cert-manager namespace. Im sure I saw an issue open about that but can't seem to find it.

**Comment by Vashiru** [comment]
> Thanks, @Vashiru! I'm glad it helped. It does only address HTTP, because the cert-manager docs have everything you need to add a DNS issuer. HTTP is universal, but DNS is specific to the provider that you want to use. Even if I made a video for one provider, it wouldn't solve everyone's needs.
> 
> Rancher 2.3 comes out next week. The cert-manager app update for the catalog is ready for merge, but it depends on a UI patch that will be in 2.3. You should be able to upgrade everything from the UI after next Tuesday.

I've upgraded to Rancher v2.3.1, but the highest cert manager I see is still just 0.6.7, is this correct? Or do I have to change a setting somewhere to see newer versions?

**Comment by vkruoso** [comment]
That seems to be the case for me too. Looks like the last chart version is not in the master branch.

**Comment by oskapt** [comment]
Yeah...there was an issue with upgrading to 0.9.1 that delayed the release. 

Honestly, because of the rate of development and the number of breaking changes, it's going to be hard to give that "one click upgrade" experience from the app catalog. There are CRDs and changes to the ClusterIssuer format, and with 0.11 they changed the annotation domain from `cert-manager.k8s.io` to `cert-manager.io` - the result of all of this is that upgrades are going to be manual and need to be done carefully.

At this point I recommend that you switch over from the app catalog version to the upstream version from Jetstack. The [video](https://youtu.be/nF_k1YZYjH0) is for 0.9.1, but the process is still the same. Just watch out for the domain change in 0.11. 

It's particularly important that you do this before November 1, because certificates from cert-manager versions < 0.8 will not renew.
